### PR TITLE
ADD EPILS

### PR DESCRIPTION
**Pull request by mlindauer**
ADD a ParamILS version with sorting neighbors by acquisition function

* side effect for SMAC: incumbent won't change if a challenger has the same performance (no walk on plateaus anymore)

I would appreciate a fast review of this PR since I need it for pSMAC

### CODE REVIEW

File: smac/facade/epils_facade.py [hunk 1]
```diff
@@ -0,0 +1,418 @@
+import logging
```
Comment (mfeurer) [inline smac/facade/epils_facade.py]: I assume a lot of code here is copied from the SMAC-facade. Maybe we should have a file utils.py in the facade directory which contains helper functions to for example retrieve the TAE in order to avoid code duplications.

File: smac/intensification/intensification.py [hunk 2]
```diff
@@ -460,10 +472,16 @@ def _compare_configs(self, incumbent: Configuration,
...
         # Line 16
         if not set(inc_runs) - set(chall_runs):
+
+            # no plateau walks
+            if chal_perf >= inc_perf:
+                self.logger.debug("Incumbent (%.4f) is better than challenger (%.4f) on %d runs." % (
+                                    inc_perf, chal_perf, len(chall_runs)))
+                return incumbent 
+            
             # Challenger is as good as incumbent
```
Comment (mfeurer) [inline smac/intensification/intensification.py]: Shouldn't this now be `Challenger is better than incumbent`?

File: smac/facade/epils_facade.py (hunk 1 above)
Comment (mlindauer) [inline smac/facade/epils_facade.py]: Yes, a lot of code was copied. I don't know whether we should go for a utils.py here or a parent class with some general functions. My intuition would be the latter.

However, I would prefer to do this in another PR and merge this one for now. Ok?

File: smac/intensification/intensification.py (hunk 2 above)
Comment (mlindauer) [inline smac/intensification/intensification.py]: Yes, thank you for finding this.

File: smac/optimizer/epils.py [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+            X, Y = self.rh2EPM.transform(self.runhistory)
+            self.model.train(X, Y)            
+            self.acquisition_func.update(model=self.model, eta=self.runhistory.get_cost(self.incumbent))
+
+
+            if iteration == 1:
+                start_point = self.incumbent
+            else:
+                # Restart?
+                if random.random() < self.restart_prob:
```
Comment (KEggensperger) [inline smac/optimizer/epils.py]: This is not seeded, right? Is that on purpose?

File: smac/optimizer/epils.py [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+            The best found configuration
+        """
+        
+        self.intensifier.minR = self.fast_race_minR # be aggressive here!
+        self.intensifier.Adaptive_Capping_Slackfactor = self.fast_race_adaptive_capping_factor
+        
+        incumbent = start_point
+        # Compute the acquisition value of the incumbent
+        incumbent_array = convert_configurations_to_array([incumbent])
+        acq_val_incumbent = self.acquisition_func(incumbent_array)
```
Comment (KEggensperger) [inline smac/optimizer/epils.py]: This is not used anywhere?

File: test/test_smbo/test_epils.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        self.assertIsInstance(epils.rng, np.random.RandomState)
+        self.assertIsInstance(epils.num_run, int)
+        epils = EPILS(self.scenario, rng=1).solver
+        rng = np.random.RandomState(1)
+        self.assertEqual(epils.num_run, 1)
+        self.assertIsInstance(epils.rng, np.random.RandomState)
+        epils = EPILS(self.scenario, rng=rng).solver
+        self.assertIsInstance(epils.num_run, int)
+        self.assertIs(epils.rng, rng)
+        # ML: I don't understand the following line and it throws an error
```
Comment (KEggensperger) [inline test/test_smbo/test_epils.py]: If this comment is still true the unittests should fail. As they don't I assume this issue is solved and the comment can be removed.

File: smac/optimizer/epils.py (hunk 3 above)
Comment (mlindauer) [inline smac/optimizer/epils.py]: thanks. fixed.

File: smac/optimizer/epils.py (hunk 4 above)
Comment (mlindauer) [inline smac/optimizer/epils.py]: correct. fixed.

File: test/test_smbo/test_epils.py (hunk 5 above)
Comment (mlindauer) [inline test/test_smbo/test_epils.py]: comment removed.

File: smac/facade/epils_facade.py (hunk 1 above)
Comment (mfeurer) [inline smac/facade/epils_facade.py]: Sure.

**Comment by mlindauer** [comment]
@mfeurer do you think we can merge this PR, too?

**Comment by mfeurer** [comment]
I think so.
